### PR TITLE
Add support for the Fixed-Size Primitive Layout

### PR DESCRIPTION
**Pull request by Benjamin-Philip**
This commit adds support for the Fixed-Size Primitive Layout:
https://arrow.apache.org/docs/format/Columnar.html#fixed-size-primitive-layout

### CODE REVIEW

**Comment by Benjamin-Philip** [comment]
As of right now the actual data is not saved as I'm yet to write a complete buffer implementation.

File: src/serde_arrow_primitive_array.erl [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    %% 2. We convert the integer bits to actual bits.
+    %%
+    %% We are doing 2 distinct operations over a single list traversal as it may
+    %% have a noticeable performance difference for large validity buffer.
+    %% We also don't generate the initial ValidityList as a binary because we
+    %% need its sum for the null count.
+
+    Bytes = lists:map(fun(Byte) -> <<<<X:1>> || X <- lists:reverse(Byte)>> end, SplitValidityList),
+    %% buffer(Bytes, byte)
+    Bytes.
```
Comment (Benjamin-Philip) [inline src/serde_arrow_primitive_array.erl:112]: I feel we could do a reduction to find the Null Count, and then do a binary comprehension to generate the validity list, and then pad and reverse the bits of each byte in binary instead of on lists.

Would this make any performance difference?

File: src/serde_arrow_array.erl [hunk 2]
```diff
@@ -0,0 +1,41 @@
+-module(serde_arrow_array).
+-export([offsets/1]).
+
+-include("serde_arrow_primitive_array.hrl").
+
+-export_type([array/0]).
+-type array() :: #primitive_array{}.
+
+-optional_callbacks([offsets/1]).
```
Comment (Benjamin-Philip) [inline src/serde_arrow_array.erl:9]: Is there a style guide regarding ordering declarations? Something similar to this [Brew Cask Documentation](https://docs.brew.sh/Cask-Cookbook#stanza-order) would suffice.

File: test/serde_arrow_primitive_array_SUITE.erl [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    Array = serde_arrow_primitive_array:new([1, 2, 3], {s, 8}),
+    ?assertEqual(serde_arrow_primitive_array:len(Array), 3).
+
+valid_null_count_on_access(_Config) ->
+    Array = serde_arrow_primitive_array:new([1, 2, 3], {s, 8}),
+    ?assertEqual(serde_arrow_primitive_array:null_count(Array), 0).
+
+valid_validity_bitmap_on_access(_Config) ->
+    Array = serde_arrow_primitive_array:new([1, 2, 3], {s, 8}),
+    ?assertEqual(serde_arrow_primitive_array:validity_bitmap(Array), nil).
```
Comment (Benjamin-Philip) [inline test/serde_arrow_primitive_array_SUITE.erl:58]: Should I just test `new/1` and use the accessor functions instead of using the record syntax, or is this fine?

File: test/serde_arrow_primitive_array_SUITE.erl [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+-include("serde_arrow_primitive_array.hrl").
+
+all() ->
+    [
+        valid_len_on_new,
+        valid_null_count_on_new,
+        valid_validity_bitmap_on_new,
+        valid_len_on_access,
+        valid_null_count_on_access,
+        valid_validity_bitmap_on_access
```
Comment (Benjamin-Philip) [inline test/serde_arrow_primitive_array_SUITE.erl:17]: I don't like these names. Is there a convention I could follow (in CommonTest tests in general)?

File: src/serde_arrow_primitive_array.erl [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+new(Value, Type) ->
+    Len = length(Value),
+    {NullCount, Bitmap} = validity(Value, Len),
+    Bin = buffer(Value, Type),
+    #primitive_array{
+        type = Type, len = Len, null_count = NullCount, validity_bitmap = Bitmap, value = Bin
+    }.
+
+%% TODO: Extract helper functions to a utils module.
+
```
Comment (Benjamin-Philip) [inline src/serde_arrow_primitive_array.erl:62]: I will extract the helper functions later when I have to implement another layout.

File: src/serde_arrow_primitive_array.erl (hunk 1 above)
Comment (Benjamin-Philip) [reply src/serde_arrow_primitive_array.erl:112]: It's quite odd that there isn't an `Enum.count/2` equivalent in `lists`. It's way more odd that there isn't a standard `reduce/2` implementation for lists. I guess if we decided to go with the reduction route, I will have to manually write the counting function.

**Comment by Benjamin-Philip** [comment]
@josevalim and @tsloughter,  Any thoughts before I write a buffer implementation? It's a bit messy at this point, and I'm procastinating writing the docs, but it should clear up soon.

File: test/serde_arrow_primitive_array_SUITE.erl (hunk 4 above)
Comment (tsloughter) [reply test/serde_arrow_primitive_array_SUITE.erl:17]: Na, no real convention. I'd probably combine these into like a test of `primitive_array_creation_tests` and just do all the tests for that in the one function.

File: src/serde_arrow_primitive_array.hrl [hunk 6]
```diff
@@ -0,0 +1,7 @@
+-record(primitive_array, {
+    type :: serde_arrow_type:arrow_type(),
+    len :: pos_integer(),
+    null_count :: pos_integer(),
+    validity_bitmap :: binary() | nil,
```
Comment (tsloughter) [inline src/serde_arrow_primitive_array.hrl:5]:
```suggestion
    validity_bitmap :: binary() | undefined,
```

`undefined` is the convention in Erlang.

File: src/serde_arrow_array.erl (hunk 2 above)
Comment (tsloughter) [reply src/serde_arrow_array.erl:9]: Not that I know of. What you have is what I tend to do (aside from I put `export_type` after all the `type` declarations).

File: src/serde_arrow_primitive_array.hrl (hunk 6 above)
Comment (Benjamin-Philip) [reply src/serde_arrow_primitive_array.hrl:5]: Ok. Quite odd that Elixir does not follow that convention. In that case would all "nil" or "null" values in Erlang be `undefined`?

File: src/serde_arrow_array_utils.erl [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    %% element such that every resulting list has a length of 8 elements.
+    %%
+    %% If we try to split a list with only 8 elements, we will get a copy of the
+    %% list and an empty list in the resulting tuple. In order to get over this
+    %% empty list, we pad one bit extra and drop the final list.
+
+    PadLen = 9 - Len rem 8,
+    Padding = lists:duplicate(PadLen, 0),
+    SplitValidityList = lists:droplast(
+        tuple_to_list(lists:split(8, lists:append(ValidityList, Padding)))
```
Comment (tsloughter) [inline src/serde_arrow_array_utils.erl:37]: If I'm reading the comment right you want to split the list into segments of 8 elements? This looks to be splitting the list into 1 element of at most 8 elements and another of the rest.

Comment (Benjamin-Philip) [reply src/serde_arrow_array_utils.erl:37]: So the initial list may have any positive number (i.e >  0) of elements. We need to pad the list so that the number of elements is a multiple of 8. We then need to split the list into segments of 8 elements, where each element is made into a byte.

We are padding an extra bit to remove the empty list that comes as `List2` as a result of splitting a list of just 8 elements. We then can remove the extra bit, which is guaranteed to be extra regardless whether we are splitting a 7 element list or a 1025 element list.

The documentation as to *why* we are doing this can be found [here](https://arrow.apache.org/docs/format/Columnar.html#validity-bitmaps).

File: src/serde_arrow_primitive_array.hrl (hunk 6 above)
Comment (Benjamin-Philip) [reply src/serde_arrow_primitive_array.hrl:5]: How will we consume lists from Elixir? Should we accept both `undefined` and `nil` as null values?

File: src/serde_arrow_array.erl (hunk 2 above)
Comment (Benjamin-Philip) [reply src/serde_arrow_array.erl:9]: Hmm.. Somebody should write one. It's nice to have some consistency with things like this as a community.

File: test/serde_arrow_primitive_array_SUITE.erl (hunk 3 above)
Comment (Benjamin-Philip) [reply test/serde_arrow_primitive_array_SUITE.erl]: I'm assuming that this okay for now... 

I'm asking because some reviewers prefer not testing accessor functions like this because unless you have don't something monumentally stupid, it's bound to work. The reason I'm testing it anyway is because I think I'm capable of doing something monumentally stupid in the future.

File: src/serde_arrow_array_utils.erl (hunk 7 above)
Comment (Benjamin-Philip) [reply src/serde_arrow_array_utils.erl:39]: I do want N / 8 lists. This does give me N / 8 lists. When I split an list of 24 elements, it give a three tuple.

Perhaps I should write a test case for this.

Comment (tsloughter) [reply src/serde_arrow_array_utils.erl:39]: Hm, then I'm missing something because `lists:split/2` returns a 2-tuple, always.

Comment (Benjamin-Philip) [reply src/serde_arrow_array_utils.erl:39]: You're right! When I tested this with a 24 element list I received incorrect output.
I'll get this fixed right study

Comment (josevalim) [reply src/serde_arrow_array_utils.erl:39]: I believe a recursive version would be clearer and more efficient. Something like this:

```elixir
-define(validity(X), if X =:= nil or X =:= undefined -> 1; true -> 0 end).

something([X1, X2, X3, X4, X5, X6, X7, X8 | Rest], Acc, Count) ->
  B1 = ?validity(X1),
  B2 = ?validity(X2),
  B3 = ?validity(X3),
  B4 = ?validity(X4),
  B5 = ?validity(X5),
  B6 = ?validity(X6),
  B7 = ?validity(X7),
  B8 = ?validity(X8),

  something(
    Rest,
    <<Acc/binary, B1:1, B2:1, B3:1, B4:1, B5:1, B6:1, B7:1, B8:1>>,
    Count + B1 + B2 + B3 + B4 + B5 + B6 + B7 + B8
  ).
something([], Acc, Count) ->
  {Acc, Count};
something(LeftOver, Acc, Count) ->
  something(pad(LeftOver), Acc, Count).
```

It is a single pass, except for the padding. No closures, no high-order functions. Please give it a try.

Comment (josevalim) [reply src/serde_arrow_array_utils.erl:39]: It is no problem, we are not expecting anyone to come up with the best solution upfront, the important is to listen, learn, and collaborate. :)

**Comment by Benjamin-Philip** [comment]
@josevalim and @tsloughter, I've written an intermediate buffer representation implementation, with which my implementation of an intermediate representation of the primitive array is complete, and this PR becomes review ready. Currently I would describe the code as brown (i.e. TDD brown) or refactorable, and I only pushed the previous commit to avoid spending energy beautifying something that may get rewritten. There are some additional things I need to do before we can merge:

1. Write a better commit message for this PR
2. Write better, more complete and detailed documentation
3. Do some potentional refactors
4. Write some more potential test cases just to be sure.

Do have a look as soon as possible as I would like to have this wrapped up before this Good Friday.

**Comment by tsloughter** [comment]
I'll get this reviewed today.

File: src/serde_arrow_array_utils.erl [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+%%    of each batch of 8 elements make up what is called the Validity Bitmap.
+%%
+%% 5. If the Null Count is 0, we can allocate the Validity Bitmap as a NULL
+%%    pointer (which in Erlang's case is `undefined').
+%%
+%% [1]: https://arrow.apache.org/docs/format/Columnar.html#validity-bitmaps
+%% [2]: https://arrow.apache.org/docs/format/Columnar.html#null-count
+
+validity_bitmap(Value) ->
+    case (lists:member(undefined, Value)) or (lists:member(nil, Value)) of
```
Comment (tsloughter) [inline src/serde_arrow_array_utils.erl:38]:
```suggestion
    case (lists:member(undefined, Value)) orelse (lists:member(nil, Value)) of
```

`orelse` and `andalso` are the short-circuit version of or/and in Erlang.

File: src/serde_arrow_array_utils.erl [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    %% Here we both pad as well as LSB number at one pass.
+    PadLen = 8 - Len rem 8,
+    Padded = lists:duplicate(PadLen, 0) ++ lists:reverse(Validities),
+
+    [B1, B2, B3, B4, B5, B6, B7, B8] = Padded,
+    bitmap([], Acc ++ [<<B1:1, B2:1, B3:1, B4:1, B5:1, B6:1, B7:1, B8:1>>], NullCount + Nulls).
+
+validity(X) ->
+    if
+        (X =:= nil) or (X =:= undefined) ->
```
Comment (tsloughter) [inline src/serde_arrow_array_utils.erl:78]:
```suggestion
        (X =:= nil) orelse (X =:= undefined) ->
```

File: src/serde_arrow_buffer.erl [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+    #buffer{type = Type, length = Len, element_length = ElementLen, data = Bin}.
+
+binary(Values, byte, Len, _ElementLen) ->
+    %% Since binary with byte will only be called to make the validity bitmap,
+    %% we do not need to account for nil values.
+    pad(<<X || X <- Values>>, pad_len(Len));
+binary(Values, Type, Len, ElementLen) ->
+    pad(<<(slot(X, Type, ElementLen)) || X <- Values>>, pad_len(Len)).
+
+slot(Value, _Type, ElementLen) when (Value =:= undefined) or (Value =:= nil) ->
```
Comment (tsloughter) [inline src/serde_arrow_buffer.erl:50]:
```suggestion
slot(Value, _Type, ElementLen) when (Value =:= undefined) orelse (Value =:= nil) ->
```

File: src/serde_arrow_type.erl [hunk 11]
```diff
@@ -51,8 +50,31 @@
...
-    | nil
+    | undefined
     | integer()
     | float().
 %% Any Erlang datatype which `serde_arrow' supports serializing from and
 %% deserializing into.
+
+-spec bit_length(Type :: arrow_type()) -> Length :: pos_integer().
+%% doc Returns the size of the type in bits.
+bit_length({Type, Size}) when (Type =:= s) or (Type =:= u) or (Type =:= f) ->
```
Comment (tsloughter) [inline src/serde_arrow_type.erl:61]:
```suggestion
bit_length({Type, Size}) when (Type =:= s) orelse (Type =:= u) orelse (Type =:= f) ->
```

**Review by tsloughter — DISMISSED** [review]
Looks good, just have some changes related to Erlang's odd `orelse` :)

File: test/serde_arrow_primitive_array_SUITE.erl (hunk 3 above)
Comment (tsloughter) [reply test/serde_arrow_primitive_array_SUITE.erl]: This is good, no reason not to test them.

File: src/serde_arrow_primitive_array.erl (hunk 1 above)
Comment (tsloughter) [reply src/serde_arrow_primitive_array.erl]: I wouldn't worry about performance.

Is `lists:foldl` not the same as `Enum.reduce`?

File: src/serde_arrow_array_utils.erl [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+    B6 = validity(X3),
+    B5 = validity(X4),
+    B4 = validity(X5),
+    B3 = validity(X6),
+    B2 = validity(X7),
+    B1 = validity(X8),
+
+    bitmap(
+        Rest,
+        Acc ++ [<<B1:1, B2:1, B3:1, B4:1, B5:1, B6:1, B7:1, B8:1>>],
```
Comment (josevalim) [inline src/serde_arrow_array_utils.erl:59]: Is there a reason for not keeping `Acc` as a binary and then doing `<<Acc/binary, B1:1, B2:1, B3:1, B4:1, B5:1, B6:1, B7:1, B8:1>>`?

If you really want to keep this as a list, then you should never append to a list on a loop, as that will copy the whole list on every appending. Instead you should do `[<<B1:1, B2:1, B3:1, B4:1, B5:1, B6:1, B7:1, B8:1>> | Acc]` (or `++ Acc`) but keeping it as a binary should be the most efficient.

File: src/serde_arrow_array_utils.erl (hunk 8 above)
Comment (Benjamin-Philip) [reply src/serde_arrow_array_utils.erl]: You didn't have to manually change every instance! You just had to tell me to change it. It was just a simple `M-% ) or RET ) orelse RET` (for a single file) or a `sed -i 's/) or/) orelse/g' src/*.erl` (for all files) for me to change it (with some possible change to the regexp)!

File: src/serde_arrow_array_utils.erl (hunk 12 above)
Comment (Benjamin-Philip) [reply src/serde_arrow_array_utils.erl:59]: The main reason for keeping it as a list was so that it was valid input for `buffer:new`. However, I had to manage the byte input exception in the buffer implementation, and it's not like to code as it is now is *elegently reusable* either, so I see no reason to not change `buffer:new`'s implementation to accept binaries.

TLDR; We'll use binaries instead.

File: src/serde_arrow_primitive_array.erl (hunk 1 above)
Comment (Benjamin-Philip) [reply src/serde_arrow_primitive_array.erl]: Resolving since we have rewritten this function.

File: src/serde_arrow_array_utils.erl (hunk 12 above)
Comment (Benjamin-Philip) [reply src/serde_arrow_array_utils.erl]: I've pushed a commit addressing and thus resolving this.

File: test/serde_arrow_type_SUITE.erl [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+    ?assertEqual(serde_arrow_type:byte_length({u, 8}), 1),
+    ?assertEqual(serde_arrow_type:byte_length({f, 32}), 4),
+
+    %% This case is stub for the current function's stub implementation
+    ?assertEqual(serde_arrow_type:byte_length(arrow_boolean), 1).
+
+serialize(_Config) ->
+    ?assertEqual(serde_arrow_type:serialize(1, {s, 64}), <<1:64/little-signed-integer>>),
+    ?assertEqual(serde_arrow_type:serialize(1, {u, 64}), <<1:64/little-unsigned-integer>>),
+    ?assertEqual(serde_arrow_type:serialize(1, {f, 64}), <<1.0:64/little-float>>).
```
Comment (Benjamin-Philip) [inline test/serde_arrow_type_SUITE.erl:33]: I do not check if the type supplied is valid or not in all these functions. Thus if I supply an incorrect type I will probably get a bad_match error. Should I match for invalid types and pass an error, or should I just leave it because this something internal and invalid types are almost impossible.

File: test/serde_arrow_buffer_SUITE.erl [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+valid_element_length_on_from_binary(_Config) ->
+    Buffer = serde_arrow_buffer:from_binary(<<1, 2, 3>>, byte, 3, 1),
+    ?assertEqual(Buffer#buffer.length, 3).
+
+valid_type_on_from_binary(_Config) ->
+    Buffer1 = serde_arrow_buffer:from_binary(<<1, 2, 3>>, byte, 3, 1),
+    ?assertEqual(Buffer1#buffer.type, byte),
+
+    Buffer2 = serde_arrow_buffer:from_binary(<<1, 2, 3>>, {s, 8}, 3, 1),
+    ?assertEqual(Buffer2#buffer.type, {s, 8}).
```
Comment (Benjamin-Philip) [inline test/serde_arrow_buffer_SUITE.erl:88]: I also do not check if the binary size and the element length supplied are valid or not. I could do `{s, 8}` with an element length of `8` bytes and it would work. Similarly, a binary of `<<1, 2, 3, 4>>` could be given a size of `0` and it would pass without any error.

The question now is if I should write checks for these, or should I just ignore this, since only I *personally* will use this function and suply values to it, which would always be valid.

The reason I ask is because adding checks just leads to more tests and code to maintain, code which wouldn't add any value 99% of the time.

**Review by Benjamin-Philip — COMMENTED** [review]
I have a few questions for adding checks to certain functions and passing errors. Please do have a look.

**Comment by Benjamin-Philip** [comment]
@josevalim I've switched to binary accumulations in the bitmap as suggested.

File: test/serde_arrow_bitmap_SUITE.erl [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+        <<1:1, 1:1, 1:1, 1:1, 1:1, 0:1, 1:1, 1:1, 0:1, 0:1, 0:1, 0:1, 0:1, 0:1, 1:1, 0:1>>
+    ),
+    ?assertEqual(Bitmap5, Buffer5).
+
+%%%%%%%%%%%
+%% Utils %%
+%%%%%%%%%%%
+
+buffer(Bitmap) ->
+    serde_arrow_buffer:from_binary(Bitmap, byte, byte_size(Bitmap), 1).
```
Comment (Benjamin-Philip) [inline test/serde_arrow_bitmap_SUITE.erl:66]: This function is copied from the Primitive Array test suite. What is the convention for defining helper functions **that will only be used in the test suites** in Common Test?  My google-fu is failing me. For example, had we been using ExUnit, this function would have gone into `SerdeArrow.Case`.

Comment (tsloughter) [reply test/serde_arrow_bitmap_SUITE.erl:66]: You can put any module in `test/`. I tend to have something like `serde_arrow_utils.erl` or `common` in there for stuff like this.

File: test/serde_arrow_type_SUITE.erl (hunk 13 above)
Comment (tsloughter) [reply test/serde_arrow_type_SUITE.erl:33]: I'd leave it.

Comment (Benjamin-Philip) [reply test/serde_arrow_type_SUITE.erl:33]: Cool.

File: test/serde_arrow_buffer_SUITE.erl (hunk 14 above)
Comment (Benjamin-Philip) [reply test/serde_arrow_buffer_SUITE.erl:88]: > I'd leave it

Same for this?

**Comment by Benjamin-Philip** [comment]
@tsloughter and @josevalim, I've made the changes requested, and some more. You can now review it - I don't plan to make anymore changes or additions. All I have to do now is write the docs in a more EDoc friendly format and probably add more docs where it is missing.

File: test/serde_arrow_buffer_SUITE.erl (hunk 14 above)
Comment (tsloughter) [reply test/serde_arrow_buffer_SUITE.erl:88]: Yea. Can always change your mind later, but less code to deal with in the mean time.

**Review by tsloughter — APPROVED** [review]

File: src/serde_arrow_type.erl [hunk 16]
```diff
@@ -21,9 +22,7 @@
     | arrow_float().
 %% Any primitive logical type in Apache Arrow that is supported by `serde_arrow'.
 
--type arrow_bool() ::
-    arrow_true
-    | arrow_false.
+-type arrow_bool() :: arrow_boolean.
```
Comment (josevalim) [inline src/serde_arrow_type.erl:25]: IMO we should just call it `boolean` or `b`. We don't call floats `arrow_floats`, we just use `f` for them.

**Review by josevalim — COMMENTED** [review]
Late review, but beautifully done. Just one tiny comment.

File: src/serde_arrow_type.erl (hunk 16 above)
Comment (Benjamin-Philip) [reply src/serde_arrow_type.erl:25]: Resolving as addressed in the next PR.
